toplevel/coqloop.ml

```diff
@@ -488,43 +506,32 @@ let read_and_execute ~state =
     let msg = CErrors.iprint (e, info) in
     TopErr.print_error_for_buffer ?loc Feedback.Error msg top_buffer;
     if exit_on_error () then exit 1;
-    state, false
-
-(* This function will only return on [Drop], careful to keep it tail-recursive *)
-let rec vernac_loop ~state =
-  let open Vernac.State in
-  loop_flush_all ();
-  top_stderr (fnl());
-  if !print_emacs then top_stderr (str (top_buffer.prompt state.doc));
-  resynch_buffer top_buffer;
-  let state, drop = read_and_execute ~state in
-  if drop then state else (vernac_loop [@ocaml.tailcall]) ~state
-
-(* Default toplevel loop, machinery for drop is below *)
-
-let drop_args = ref None
-
-(* Initialises the Ocaml toplevel before launching it, so that it can
-   find the "include" file in the *source* directory *)
-let init_ocaml_path () =
-  let env = Boot.Env.init () in
-  let corelib = Boot.Env.corelib env |> Boot.Path.to_string in
-  let add_subdir dl = Mltop.add_ml_dir (Filename.concat corelib dl) in
-  List.iter add_subdir ("dev" :: Coq_config.all_src_dirs)
+    true, state
 
-let loop ~opts ~state =
-  drop_args := Some opts;
+let loop ~state =
+  (* Initialize buffer *)
+  reset_input_buffer ~state;
+  Flags.without_option Flags.in_ml_toplevel (fun () ->
+    (* The main loop *)
+    let running = ref true in
+    let state = ref state in
+    while !running do
```

Comment:
Style in this file is to use tail-recursion and no refs, please rewrite.

Rest looks good to me, thanks!